### ISSUE TITLE
ISwaggerProvider can't be requested multiple times

### ISSUE THREAD

**Issue by 0x53A** [domaindrivendev/Swashbuckle.AspNetCore#791]
This code works on the first request, and fails on the second request:

```C#
                               var svc1 = ctx.RequestServices.GetRequiredService<ISwaggerProvider>();
                               var doc1 = svc1.GetSwagger(doc, null, ctx.Request.PathBase);
```

This fails on the first request on line ``doc2 =``:

```C#
                               var svc1 = ctx.RequestServices.GetRequiredService<ISwaggerProvider>();
                               var doc1 = svc1.GetSwagger(doc, null, ctx.Request.PathBase);

                               var svc2 = ctx.RequestServices.GetRequiredService<ISwaggerProvider>();
                               var doc2 = svc1.GetSwagger(doc, null, ctx.Request.PathBase);
```

> System.ArgumentException: 'Ein Element mit dem gleichen Schlüssel wurde bereits hinzugefügt.' (An element with the same key was already added)

```
 	mscorlib.dll!System.ThrowHelper.ThrowArgumentException(System.ExceptionResource resource)	Unknown
 	mscorlib.dll!System.Collections.Generic.Dictionary<string, Swashbuckle.AspNetCore.Swagger.Header>.Insert(string key, Swashbuckle.AspNetCore.Swagger.Header value, bool add)	Unknown
 	mscorlib.dll!System.Collections.Generic.Dictionary<System.__Canon, System.__Canon>.Add(System.__Canon key, System.__Canon value)	Unknown
 	Swashbuckle.AspNetCore.Examples.dll!Swashbuckle.AspNetCore.Examples.AddResponseHeadersFilter.Apply(Swashbuckle.AspNetCore.Swagger.Operation operation, Swashbuckle.AspNetCore.SwaggerGen.OperationFilterContext context)	Unknown
 	Swashbuckle.AspNetCore.SwaggerGen.dll!Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreateOperation(Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription apiDescription, Swashbuckle.AspNetCore.SwaggerGen.ISchemaRegistry schemaRegistry)	Unknown
 	Swashbuckle.AspNetCore.SwaggerGen.dll!Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreatePathItem(System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription> apiDescriptions, Swashbuckle.AspNetCore.SwaggerGen.ISchemaRegistry schemaRegistry)	Unknown
 	Swashbuckle.AspNetCore.SwaggerGen.dll!Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreatePathItems.AnonymousMethod__2(System.Linq.IGrouping<string, Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription> group)	Unknown
 	System.Core.dll!System.Linq.Enumerable.ToDictionary<System.Linq.IGrouping<string, Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription>, string, Swashbuckle.AspNetCore.Swagger.PathItem>(System.Collections.Generic.IEnumerable<System.Linq.IGrouping<string, Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription>> source, System.Func<System.Linq.IGrouping<string, Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription>, string> keySelector, System.Func<System.Linq.IGrouping<string, Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription>, Swashbuckle.AspNetCore.Swagger.PathItem> elementSelector, System.Collections.Generic.IEqualityComparer<string> comparer)	Unknown
 	Swashbuckle.AspNetCore.SwaggerGen.dll!Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreatePathItems(System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescription> apiDescriptions, Swashbuckle.AspNetCore.SwaggerGen.ISchemaRegistry schemaRegistry)	Unknown
 	Swashbuckle.AspNetCore.SwaggerGen.dll!Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GetSwagger(string documentName, string host, string basePath, string[] schemes)	Unknown

```

![grafik](https://user-images.githubusercontent.com/4236651/42330349-49e904cc-8073-11e8-83e8-0e6ca3a15f26.png)

I *think* the issue is that the service is registered as transient, which means that asp.net core will create a new instance every time, but the initializer function will (each time) execute all the filters.

Replacing that call with AddSingleton should fix the issue.

**Comment by domaindrivendev** [comment]
Please see #629. This was a previous PR to make ISwaggerProvider singleton. However, I determined that the failure on transient service lifetime was a symptom of the problem rather than the cause. The root cause is incorrect handling state in the Swashbuckle plumbing. I'm about to check in a fix for this

**Comment by domaindrivendev** [comment]
Fixed by d5fa89b7c675a52a38745bdab4ffa3e79a5f7d6e. Will be released with 3.0.0 (coming soon). In the meantime you can pull the latest preview from myget.org and try it out.

**Comment by 0x53A** [comment]
Thank you. I've worked around this issue by caching the service, like the MiddleWare does, so there is no rush from my side.
